### PR TITLE
DECODER: Use the data structure from the last decoder to select the next

### LINKED ISSUE #13539: How to load a provider-side RSA key using the default file STORE
I have troubles loading a provider-side RSA key using the default file STORE implementation and I wonder what is the right way implementing this. I did the following:
 - In my provider both KEYMGMT and ENCODER/DECODER algorithms are named "RSA" because this is RSA (just TPM-based) and I want these to work (out of a box) with the standard openssl tools and libraries.
 - I wanted to re-use the "file" STORE, so my OSSL_provider_init() loads also the default provider and my OSSL_FUNC_PROVIDER_QUERY_OPERATION returns the default OSSL_OP_STORE algorithm.

This does not work as expected though:
 - When I call e.g. `openssl pkey -provider tpm2 -in testkey -pubout` the default file STORE fetched and works as expected, my DECODER is called and returns a custom testkey structure. So far so good.
 - Now, in store_result.c / ossl_store_handle_load_result() is the (RSA) KEYMGMT fetched using the STORE's context (since I reuse default provider). The properties used for fetching are NULL and (thus) the default (RSA) KEYMGMT implementation is loaded, not mine. This obviously causes a crash because the decoded key I am loading expects a different KEYMGMT.

Is there a way out of this, please? Could I reuse the STORE but still load a key into my KEYMGMT named "RSA"? The internal providers solve it by having an own STORE implementation that reuses the same ossl_file_store_functions. Unfortunately this cannot be done in external providers because these functions are internal only.

### PR DESCRIPTION
**Pull request by levitte**
Any decoder can now also declare the name of the data structure for
the object it decoded in the OSSL_PARAM array they pass back to the
decoding process.  The decoding process will use that as another
criterion to select the next decoder in the chain to consider.

Together with declaring the data type, this becomes a means to refine
how the decoded data is treated along the chain.

Fixes #13539

### CODE REVIEW

**Comment by levitte** [comment]
It's possible that this will resolve #14192 as well.

**Comment by levitte** [comment]
This is almost ready for review.  @gotthardp, if you would try my suggestion (https://github.com/openssl/openssl/issues/13539#issuecomment-781903958) against this PR, I'd appreciate it.

**Comment by levitte** [comment]
As far as I can see, the fuzzing failure isn't relevant

Review, please!

File: crypto/encode_decode/decoder_lib.c [hunk 1]
```diff
@@ -632,12 +734,20 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
...
+                       "(ctx %p) %s [%u] Running decoder instance %p => %d\n",
+                       (void *)new_data.ctx, LEVEL, (unsigned int)i,
+                       (void *)new_decoder_inst, ok);
         } OSSL_TRACE_END(DECODER);
 
         if (ok)
             break;
+
+        /*
+         * These are errors are assumed to come from
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:746]: These errors are?

File: crypto/encode_decode/decoder_lib.c [hunk 2]
```diff
@@ -647,10 +757,11 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
...
             || (ERR_GET_LIB(err) == ERR_LIB_X509
                 && ERR_GET_REASON(err) == X509_R_UNSUPPORTED_ALGORITHM))
-            break; /* fatal error; preserve it on the error queue and stop */
+            goto end;
     }
 
  end:
     BIO_free(new_data.bio);
     return ok;
+
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:766]: Why the spurious blank line?

File: crypto/encode_decode/decoder_lib.c [hunk 3]
```diff
@@ -557,8 +571,40 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
 
         /* Get the object type if there is one */
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:572]: Should this refer to data type here?

File: crypto/encode_decode/decoder_lib.c [hunk 4]
```diff
@@ -557,8 +571,40 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
...
         p = OSSL_PARAM_locate_const(params, OSSL_OBJECT_PARAM_DATA_TYPE);
-        if (p != NULL && !OSSL_PARAM_get_utf8_string_ptr(p, &object_type))
+        if (p != NULL && !OSSL_PARAM_get_utf8_string_ptr(p, &data_type))
             goto end;
+
+        /* Get the object structure if there is one */
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:577]: data structure?

File: crypto/encode_decode/decoder_lib.c [hunk 5]
```diff
@@ -512,20 +514,32 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
...
-    const char *object_type = NULL;
+    const char *data_type = NULL;
+    const char *data_structure = NULL;
 
     memset(&new_data, 0, sizeof(new_data));
     new_data.ctx = data->ctx;
+    new_data.recursion = data->recursion + 1;
+
+#define LEVEL_STR ">>>>>>>>>>>>>>>>"
+#define LEVEL (&LEVEL_STR[sizeof(LEVEL_STR) - new_data.recursion - 1])
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:525]: I know this is only relevant if we are tracing - but shouldn't we handle the case where `new_data.recursion >= sizeof(LEVEL_STR)`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/encode_decode/decoder_lib.c (hunk 5 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:525]: There is a limit to the amount of levels: 10

File: crypto/encode_decode/decoder_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:572]: Ah, yes it should!

File: crypto/encode_decode/decoder_lib.c (hunk 4 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:577]: Yup

File: crypto/encode_decode/decoder_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:746]: ... apparent in the code following the comment :wink:

File: crypto/encode_decode/decoder_lib.c (hunk 2 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:766]: "oops" is a good reason, innit? 😉 
I'll remove it

File: crypto/encode_decode/decoder_lib.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/encode_decode/decoder_lib.c:746]: I was intending to comment on the grammar..shouldn't this be "These errors are assumed..."

Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:746]: Oh!  My eyes failed me.  Ok, will fix

File: crypto/encode_decode/decoder_lib.c (hunk 5 above)
Comment (mattcaswell) [reply crypto/encode_decode/decoder_lib.c:525]: Where is that implemented? Perhaps a comment here?

Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:525]: Hmmmm, it is of course possible for the caller to add decoders at will.  So ok, I recoded the macro for the case you mention.

File: crypto/encode_decode/decoder_lib.c [hunk 6]
```diff
@@ -522,7 +522,9 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
     new_data.recursion = data->recursion + 1;
 
 #define LEVEL_STR ">>>>>>>>>>>>>>>>"
-#define LEVEL (&LEVEL_STR[sizeof(LEVEL_STR) - new_data.recursion - 1])
+#define LEVEL (new_data.recursion < sizeof(LEVEL_STR) - 1               \
```
Comment (mattcaswell) [inline crypto/encode_decode/decoder_lib.c:525]: Couldn't this be <= here?

File: crypto/encode_decode/decoder_lib.c [hunk 7]
```diff
@@ -582,31 +630,80 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
...
+                           "(ctx %p) %s [%u] the input type doesn't match the name of the previous decoder (%p), skipping...\n",
+                           (void *)new_data.ctx, LEVEL, (unsigned int)i,
+                           (void *)decoder);
+            } OSSL_TRACE_END(DECODER);
             continue;
+        }
 
         /*
-         * If the previous decoder gave us an object type, we check to see
+         * If the previous decoder gave us an data type, we check to see
```
Comment (paulidale) [inline crypto/encode_decode/decoder_lib.c:680]: `s/ an / a /`

File: crypto/encode_decode/decoder_lib.c (hunk 6 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c]: It could...  or I could remove `- 1`
Done

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/encode_decode/decoder_lib.c (hunk 7 above)
Comment (levitte) [reply crypto/encode_decode/decoder_lib.c:680]: Fixed

**Review by mattcaswell — APPROVED** [review]
LGTM

**Review by paulidale — APPROVED** [review]
Looks okay.

I do worry about the complexity of chaining decoders together but that's for another discussion.

**Comment by levitte** [comment]
Additionally, @gotthardp just gave his thumbs up: https://github.com/openssl/openssl/issues/13539#issuecomment-782117371

**Comment by levitte** [comment]
Merged

acf497b53b0a349af13ca5e89665f331e1096af8 DECODER: Use the data structure from the last decoder to select the next
57acc56bdcdf2a7f084cf480f6f1d8f250735b0c DECODER: Add better tracing of the chain walking process

**Pull request closed without merging**